pkg/storage/unified/resource/access.go

```diff
@@ -33,104 +37,153 @@ var _ authz.AccessClient = &staticAuthzClient{}
+		span.SetAttributes(attribute.String("error", err.Error()))
 		return resp, err
 	}
-	c.logger.Debug("Check", "group", req.Group, "resource", req.Resource, "fallback", false, "rbac", true, "allowed", resp.Allowed, "duration", time.Since(t))
+	span.SetAttributes(attribute.Bool("allowed", resp.Allowed))
+	c.metrics.checkDuration.WithLabelValues(req.Group, req.Resource, req.Verb, fmt.Sprintf("%t", resp.Allowed)).Observe(time.Since(t).Seconds())
 	return resp, nil
 }
 
 // Compile implements authz.AccessClient.
 func (c authzLimitedClient) Compile(ctx context.Context, id claims.AuthInfo, req authz.ListRequest) (authz.ItemChecker, error) {
+	t := time.Now()
+	fallbackUsed := grpcutils.FallbackUsed(ctx)
 	ctx, span := c.tracer.Start(ctx, "authzLimitedClient.Compile", trace.WithAttributes(
 		attribute.String("group", req.Group),
 		attribute.String("resource", req.Resource),
+		attribute.String("namespace", req.Namespace),
+		attribute.String("verb", req.Verb),
+		attribute.Bool("fallback_used", fallbackUsed),
 	))
 	defer span.End()
-	return func(namespace string, name, folder string) bool {
-		ctx, span := c.tracer.Start(ctx, "authzLimitedClient.Compile.Check", trace.WithAttributes(
-			attribute.String("group", req.Group),
-			attribute.String("resource", req.Resource),
-			attribute.Bool("fallback", grpcutils.FallbackUsed(ctx)),
-		))
-		defer span.End()
-		if grpcutils.FallbackUsed(ctx) {
-			c.logger.Debug("Compile.Check", "group", req.Group, "resource", req.Resource, "fallback", true, "rbac", false, "allowed", true)
+	if fallbackUsed || !c.IsCompatibleWithRBAC(req.Group, req.Resource) {
+		return func(namespace string, name, folder string) bool {
 			return true
-		}
-		// TODO: Implement For now we perform the check for each item.
-		if !c.IsCompatibleWithRBAC(req.Group, req.Resource) {
-			c.logger.Debug("Compile.Check", "group", req.Group, "resource", req.Resource, "namespace", namespace, "name", name, "folder", folder, "fallback", false, "rbac", false, "allowed", true)
-			return true
-		}
-		t := time.Now()
-		r, err := c.client.Check(ctx, id, authz.CheckRequest{
-			Verb:      "get",
-			Group:     req.Group,
-			Resource:  req.Resource,
-			Namespace: namespace,
-			Name:      name,
-			Folder:    folder,
-		})
-		if err != nil {
-			c.logger.Error("Compile.Check", "group", req.Group, "resource", req.Resource, "namespace", namespace, "name", name, "folder", folder, "fallback", false, "rbac", true, "error", err, "duration", time.Since(t))
-			return false
-		}
-		c.logger.Debug("Compile.Check", "group", req.Group, "resource", req.Resource, "namespace", namespace, "name", name, "folder", folder, "fallback", false, "rbac", true, "allowed", r.Allowed, "duration", time.Since(t))
-		return r.Allowed
-	}, nil
+		}, nil
+	}
+	checker, err := c.client.Compile(ctx, id, req)
+	if err != nil {
+		c.logger.Error("Compile", "group", req.Group, "resource", req.Resource, "error", err, "traceid", tracing.TraceIDFromContext(ctx, false))
```

Comment:
I wonder if `c.logger.FromContext(ctx)` does not handle tracing out of the box.

Reply:
FYI... the logger https://github.com/grafana/grafana-app-sdk/tree/main/logging does that!